File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/Indicators/TriangleIndicator.java
package com.github.anastr.speedviewlib.components.Indicators;

import android.content.Context;
import android.graphics.BlurMaskFilter;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Path;
import android.graphics.Shader;

/**
 * this Library build By Anas Altair
 * see it on <a href="https://github.com/anastr/SpeedView">GitHub</a>
 */
public class TriangleIndicator extends Indicator<TriangleIndicator> {

    private Path indicatorPath = new Path();
    private float indicatorTop = 0;

    public TriangleIndicator(Context context) {
        super(context);
        updateIndicator();
    }

    @Override
    protected float getDefaultIndicatorWidth() {
        return dpTOpx(25f);
    }

    @Override
    public float getTop() {
        return indicatorTop;
    }

    @Override
    public float getBottom() {
        return indicatorTop + getIndicatorWidth();
    }

    @Override
    public void draw(Canvas canvas, float degree) {
        canvas.save();
        canvas.rotate(90f + degree, getCenterX(), getCenterY());
        canvas.drawPath(indicatorPath, indicatorPaint);
        canvas.restore();
    }

    @Override
    protected void updateIndicator() {
        indicatorPath = new Path();
        indicatorTop = getPadding() + getSpeedometerWidth() + dpTOpx(5);
        indicatorPath.moveTo(getCenterX(), indicatorTop);
        indicatorPath.lineTo(getCenterX() - getIndicatorWidth(), indicatorTop + getIndicatorWidth());
        indicatorPath.lineTo(getCenterX() + getIndicatorWidth(), indicatorTop + getIndicatorWidth());
        indicatorPath.moveTo(0f, 0f);

        int endColor = Color.argb(0, Color.red(getIndicatorColor()), Color.green(getIndicatorColor())
                , Color.blue(getIndicatorColor()));
        Shader linearGradient = new LinearGradient(getCenterX(), indicatorTop, getCenterX(), indicatorTop + getIndicatorWidth()
                , getIndicatorColor(), endColor, Shader.TileMode.CLAMP);
        indicatorPaint.setShader(linearGradient);
    }

    @Override
    protected void setWithEffects(boolean withEffects) {
        if (withEffects && !isInEditMode()) {
            indicatorPaint.setMaskFilter(new BlurMaskFilter(15, BlurMaskFilter.Blur.SOLID));
        }
        else {
            indicatorPaint.setMaskFilter(null);
        }
    }
}

File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/indicators/TriangleIndicator.kt
package com.github.anastr.speedviewlib.components.indicators

import android.content.Context
import android.graphics.*

/**
 * this Library build By Anas Altair
 * see it on [GitHub](https://github.com/anastr/SpeedView)
 */
class TriangleIndicator(context: Context) : Indicator<TriangleIndicator>(context) {

    private var indicatorPath = Path()
    private var indicatorTop = 0f

    init {
        width = dpTOpx(25f)
    }

    override fun getTop(): Float {
        return indicatorTop
    }

    override fun getBottom(): Float {
        return indicatorTop + width
    }

    override fun draw(canvas: Canvas) {
        canvas.drawPath(indicatorPath, indicatorPaint)
    }

    override fun updateIndicator() {
        indicatorPath = Path()
        indicatorTop = speedometer!!.padding.toFloat() + speedometer!!.speedometerWidth + dpTOpx(5f)
        indicatorPath.moveTo(getCenterX(), indicatorTop)
        indicatorPath.lineTo(getCenterX() - width, indicatorTop + width)
        indicatorPath.lineTo(getCenterX() + width, indicatorTop + width)
        indicatorPath.moveTo(0f, 0f)

        val endColor = Color.argb(0, Color.red(color), Color.green(color), Color.blue(color))
        val linearGradient = LinearGradient(getCenterX(), indicatorTop, getCenterX(), indicatorTop + width, color, endColor, Shader.TileMode.CLAMP)
        indicatorPaint.shader = linearGradient
    }

    override fun setWithEffects(withEffects: Boolean) {
        if (withEffects && !speedometer!!.isInEditMode) {
            indicatorPaint.maskFilter = BlurMaskFilter(15f, BlurMaskFilter.Blur.SOLID)
        } else {
            indicatorPaint.maskFilter = null
        }
    }
}
